src/main/java/org/gridsuite/directory/server/DirectoryService.java

```diff
@@ -398,4 +408,68 @@ private boolean isPrivateForNotification(UUID parentDirectoryUuid, boolean isCur
+        return insertElement(elementAttributes, parentDirectoryUuid).flatMap(elementAttributes1 -> {
+            emitDirectoryChanged(parentDirectoryUuid, userId, isPrivateDirectory(parentDirectoryUuid), false, NotificationType.UPDATE_DIRECTORY);
+            return actionsService.insertFiltersContingencyList(elementAttributes1.getElementUuid(), content)
+                .doOnError(err -> {
+                    deleteElement(elementAttributes1.getElementUuid(), userId);
+                    emitDirectoryChanged(parentDirectoryUuid, userId, isPrivateDirectory(parentDirectoryUuid), false, NotificationType.UPDATE_DIRECTORY);
+                });
+        });
+    }
+
+    public Mono<Void> newScriptFromFiltersContingencyList(UUID id, String scriptName, String userId, UUID parentDirectoryUuid) {
+        return getElementInfos(id).flatMap(elementAttributes -> {
+            if (elementAttributes.getType() != ElementType.FILTERS_CONTINGENCY_LIST) {
+                return Mono.error(new DirectoryException(NOT_ALLOWED));
+            }
+            ElementAttributes newElementAttributes = new ElementAttributes(null, scriptName,
+                ElementType.SCRIPT_CONTINGENCY_LIST, new AccessRightsAttributes(elementAttributes.getAccessRights().isPrivate()), userId);
+
+            return insertElement(newElementAttributes, parentDirectoryUuid).flatMap(elementAttributes1 -> {
+                emitDirectoryChanged(parentDirectoryUuid, userId, isPrivateDirectory(parentDirectoryUuid), false, NotificationType.UPDATE_DIRECTORY);
+                return actionsService.newScriptFromFiltersContingencyList(id, scriptName, elementAttributes1.getElementUuid())
+                    .doOnError(err -> {
+                        deleteElement(elementAttributes1.getElementUuid(), userId);
+                        emitDirectoryChanged(parentDirectoryUuid, userId, isPrivateDirectory(parentDirectoryUuid), false, NotificationType.UPDATE_DIRECTORY);
+                    });
+            });
+        });
+    }
+
+    public Mono<Void> replaceFilterContingencyListWithScript(UUID id, String userId, UUID parentDirectoryUuid) {
+        return getElementInfos(id).flatMap(elementAttributes -> {
+            if (!userId.equals(elementAttributes.getOwner())) {
+                return Mono.error(new DirectoryException(NOT_ALLOWED));
+            }
+            if (elementAttributes.getType() != ElementType.FILTERS_CONTINGENCY_LIST) {
+                return Mono.error(new DirectoryException(NOT_ALLOWED));
+            }
+            return actionsService.replaceFilterContingencyListWithScript(id)
+                .doOnSuccess(unused -> {
+                    directoryElementRepository.updateElementType(id, ElementType.SCRIPT_CONTINGENCY_LIST.name());
+                    emitDirectoryChanged(parentDirectoryUuid, userId, isPrivateDirectory(parentDirectoryUuid), false,
+                        NotificationType.UPDATE_DIRECTORY);
+                });
+        });
+    }
```

Comment:
Shouldn't all this be in the ContingencyListService.java ? Since you created a class for that why put only the http calls on the other class

Reply:
Difficult to do without avoiding cyclic dependencies problem between DirectoryService and ContingecyListService :

DirectoryService needs ContingencyListService in 'renameElement' and 'deleteObject' methods
If we move methods 'createScriptContingencyList', 'createFiltersContingencyList', 'newScriptFromFiltersContingencyList' and 'replaceFilterContingencyListWithScript' from DirectoryService to ContingencyListService, we will need DirectoryService in ContingencyListService ....

Reply:
good point :)